File: storage/channel.go
package storage

import (
	"strings"
	"sync"
)

type ChannelStore struct {
	users    map[string]map[string][]*ChannelStoreUser
	userLock sync.Mutex

	topic     map[string]map[string]string
	topicLock sync.Mutex
}

const userModePrefixes = "~&@%+"
const userModeChars = "qaohv"

type ChannelStoreUser struct {
	nick   string
	modes  string
	prefix string
}

func NewChannelStoreUser(nick string) *ChannelStoreUser {
	user := &ChannelStoreUser{nick: nick}

	if i := strings.IndexAny(nick, userModePrefixes); i == 0 {
		i = strings.Index(userModePrefixes, string(nick[0]))
		user.modes = string(userModeChars[i])
		user.nick = nick[1:]
		user.updatePrefix()
	}

	return user
}

func (c *ChannelStoreUser) String() string {
	return c.prefix + c.nick
}

func (c *ChannelStoreUser) addModes(modes string) {
	for _, mode := range modes {
		if strings.Contains(c.modes, string(mode)) {
			continue
		}
		c.modes += string(mode)
	}
	c.updatePrefix()
}

func (c *ChannelStoreUser) removeModes(modes string) {
	for _, mode := range modes {
		c.modes = strings.Replace(c.modes, string(mode), "", 1)
	}
	c.updatePrefix()
}

func (c *ChannelStoreUser) updatePrefix() {
	for i, mode := range userModeChars {
		if strings.Contains(c.modes, string(mode)) {
			c.prefix = string(userModePrefixes[i])
			return
		}
	}
	c.prefix = ""
}

func NewChannelStore() *ChannelStore {
	return &ChannelStore{
		users: make(map[string]map[string][]*ChannelStoreUser),
		topic: make(map[string]map[string]string),
	}
}

func (c *ChannelStore) GetUsers(server, channel string) []string {
	c.userLock.Lock()

	users := make([]string, len(c.users[server][channel]))
	for i, user := range c.users[server][channel] {
		users[i] = user.String()
	}

	c.userLock.Unlock()

	return users
}

func (c *ChannelStore) SetUsers(users []string, server, channel string) {
	c.userLock.Lock()

	if _, ok := c.users[server]; !ok {
		c.users[server] = make(map[string][]*ChannelStoreUser)
	}

	c.users[server][channel] = make([]*ChannelStoreUser, len(users))
	for i, nick := range users {
		c.users[server][channel][i] = NewChannelStoreUser(nick)
	}

	c.userLock.Unlock()
}

func (c *ChannelStore) AddUser(user, server, channel string) {
	c.userLock.Lock()

	if _, ok := c.users[server]; !ok {
		c.users[server] = make(map[string][]*ChannelStoreUser)
	}

	if users, ok := c.users[server][channel]; ok {
		for _, u := range users {
			if u.nick == user {
				c.userLock.Unlock()
				return
			}
		}

		c.users[server][channel] = append(users, NewChannelStoreUser(user))
	} else {
		c.users[server][channel] = []*ChannelStoreUser{NewChannelStoreUser(user)}
	}

	c.userLock.Unlock()
}

func (c *ChannelStore) RemoveUser(user, server, channel string) {
	c.userLock.Lock()
	c.removeUser(user, server, channel)
	c.userLock.Unlock()
}

func (c *ChannelStore) RemoveUserAll(user, server string) {
	c.userLock.Lock()

	for channel := range c.users[server] {
		c.removeUser(user, server, channel)
	}

	c.userLock.Unlock()
}

func (c *ChannelStore) RenameUser(oldNick, newNick, server string) {
	c.userLock.Lock()
	c.renameAll(server, oldNick, newNick)
	c.userLock.Unlock()
}

func (c *ChannelStore) SetMode(server, channel, user, add, remove string) {
	c.userLock.Lock()

	for _, u := range c.users[server][channel] {
		if u.nick == user {
			u.addModes(add)
			u.removeModes(remove)

			c.userLock.Unlock()
			return
		}
	}

	c.userLock.Unlock()
}

func (c *ChannelStore) GetTopic(server, channel string) string {
	c.topicLock.Lock()
	topic := c.topic[server][channel]
	c.topicLock.Unlock()
	return topic
}

func (c *ChannelStore) SetTopic(topic, server, channel string) {
	c.topicLock.Lock()

	if _, ok := c.topic[server]; !ok {
		c.topic[server] = make(map[string]string)
	}

	c.topic[server][channel] = topic
	c.topicLock.Unlock()
}

func (c *ChannelStore) rename(server, channel, oldNick, newNick string) {
	for _, user := range c.users[server][channel] {
		if user.nick == oldNick {
			user.nick = newNick
			return
		}
	}
}

func (c *ChannelStore) renameAll(server, oldNick, newNick string) {
	for channel := range c.users[server] {
		c.rename(server, channel, oldNick, newNick)
	}
}

func (c *ChannelStore) removeUser(user, server, channel string) {
	for i, u := range c.users[server][channel] {
		if u.nick == user {
			users := c.users[server][channel]
			c.users[server][channel] = append(users[:i], users[i+1:]...)
			return
		}
	}
}

File: client/src/js/stores/channel.js
var Reflux = require('reflux');
var Immutable = require('immutable');
var _ = require('lodash');

var actions = require('../actions/channel');
var serverActions = require('../actions/server');

var channels = Immutable.Map();
var empty = Immutable.List();

var User = Immutable.Record({
	nick: null,
	renderName: null,
	mode: ''
});

function createUser(nick, mode) {
	return updateRenderName(new User({
		nick: nick,
		renderName: nick,
		mode: mode || ''
	}));
}

function loadUser(nick) {
	var mode;

	if (nick[0] === '@') {
		mode = 'o';
	} else if (nick[0] === '+') {
		mode = 'v';
	}

	if (mode) {
		nick = nick.slice(1);
	}

	return createUser(nick, mode);
}

function updateRenderName(user) {
	var name = user.nick;

	if (user.mode.indexOf('o') !== -1) {
		name = '@' + name;
	} else if (user.mode.indexOf('v') !== -1) {
		name = '+' + name;
	}

	return user.set('renderName', name);
}

function sortUsers(a, b) {
	a = a.renderName.toLowerCase();
	b = b.renderName.toLowerCase();

	if (a[0] === '@' && b[0] !== '@') {
		return -1;
	}
	if (b[0] === '@' && a[0] !== '@') {
		return 1;
	}
	if (a[0] === '+' && b[0] !== '+') {
		return -1;
	}
	if (b[0] === '+' && a[0] !== '+') {
		return 1;
	}
	if (a < b) {
		return -1;
	}
	if (a > b) {
		return 1;
	}
	return 0;
}

var channelStore = Reflux.createStore({
	init() {
		this.listenToMany(actions);
		this.listenTo(serverActions.connect, 'addServer');
		this.listenTo(serverActions.disconnect, 'removeServer');
		this.listenTo(serverActions.load, 'loadServers');
	},

	part(partChannels, server) {
		_.each(partChannels, function(channel) {
			channels = channels.deleteIn([server, channel]);
		});
		this.trigger(channels);
	},

	addUser(user, server, channel) {
		channels = channels.updateIn([server, channel, 'users'], empty, users => {
			return users.push(createUser(user)).sort(sortUsers);
		});
		this.trigger(channels);
	},

	removeUser(user, server, channel) {
		if (channels.hasIn([server, channel])) {
			channels = channels.updateIn([server, channel, 'users'], users => users.filter(u => u.nick !== user));
			this.trigger(channels);
		}
	},

	removeUserAll(user, server) {
		channels.get(server).forEach((v, k) => {
			channels = channels.updateIn([server, k, 'users'], users => users.filter(u => u.nick !== user));
		});
		this.trigger(channels);
	},

	renameUser(oldNick, newNick, server) {
		channels.get(server).forEach((v, k) => {
			channels = channels.updateIn([server, k, 'users'], users => {
				var i = users.findIndex(user => user.nick === oldNick);
				return users.update(i, user => {
					return updateRenderName(user.set('nick', newNick));
				}).sort(sortUsers);
			});
		});
		this.trigger(channels);
	},

	setUsers(users, server, channel) {
		users = _.map(users, user => loadUser(user)).sort(sortUsers);
		channels = channels.setIn([server, channel, 'users'], Immutable.List(users));
		this.trigger(channels);
	},

	setTopic(topic, server, channel) {
		channels = channels.setIn([server, channel, 'topic'], topic);
		this.trigger(channels);
	},

	setMode(mode) {
		var i = channels.getIn([mode.server, mode.channel, 'users']).findIndex(u => u.nick === mode.user);

		channels = channels.updateIn([mode.server, mode.channel, 'users', i], user => {
			_.each(mode.remove, function(mode) {
				user = user.set('mode', user.mode.replace(mode, ''));
			});
			
			user = user.set('mode', user.mode + mode.add);

			return updateRenderName(user);
		});
		channels = channels.updateIn([mode.server, mode.channel, 'users'], users => users.sort(sortUsers));
		this.trigger(channels);
	},

	load(storedChannels) {
		_.each(storedChannels, function(channel) {
			channels = channels.setIn([channel.server, channel.name], Immutable.Map({
				users: Immutable.List(),
				topic: channel.topic
			}));
		});
		this.trigger(channels);
	},

	addServer(server) {
		if (!channels.has(server)) {
			channels = channels.set(server, Immutable.Map());
			this.trigger(channels);
		}
	},

	removeServer(server) {
		channels = channels.delete(server);
		this.trigger(channels);
	},

	loadServers(storedServers) {
		_.each(storedServers, function(server) {
			if (!channels.has(server.address)) {
				channels = channels.set(server.address, Immutable.Map());
			}
		});
		this.trigger(channels);
	},

	getChannels(server) {
		return channels.get(server);
	},

	getUsers(server, channel) {
		return channels.getIn([server, channel, 'users']) || empty;
	},

	getTopic(server, channel) {
		return channels.getIn([server, channel, 'topic']);
	},

	getState() {
		return channels;
	}
});

module.exports = channelStore;